### ISSUE TITLE
Presence/absence of MHV or SSOe shared cookie should not interfere with one another

### ISSUE THREAD

**Issue by patrickvinograd** [department-of-veterans-affairs/va.gov-team#7127]
TK

**Comment by patrickvinograd** [comment]
the authentication_and_sso  concern had logic around figuring out if a user should be logged in or logged out based on various combinations of VA.gov and MHV cookie (and in the future, IAM cookie).

We want to make sure that in the interval while we are supporting the MHV and IAM cookies, that they don't conflict with one another:

- Right now, absence of MHV shared cookie indicates the user logged out from MHV, and therefore should be logged out of VA.gov. When a user has an SSOe based session, the lack of MHV cookie provides no signal. 

- Eventually we will implement SSOe auto-logout, where lack of an IAM cookie indicates that we should end the corresponding  VA.gov session. But likewise, if a user was not signed in using SSOe in the first place, the lack of IAM cookie provides no signal. 

Basically we need to figure out if this is an SSOe user or not, and cleanly delineate the external logout  behavior.

**Comment by patrickvinograd** [comment]
Might be worth putting together a quick state diagram describing the  current/interim/final behavior w.r.t. cookies.

**Comment by ericbuckley** [comment]
@patrickvinograd I've been going over the authentication and sso concerns module today and still don't quite understand how it relates to this bullet point from above

> Eventually we will implement SSOe auto-logout, where lack of an IAM cookie indicates that we should end the corresponding VA.gov session. But likewise, if a user was not signed in using SSOe in the first place, the lack of IAM cookie provides no signal.

Our SSOe auto-logout procedure will be initiated from the FE, as its the only place where we can accurately test if an active SSOe session exists.  So likewise I would think that the FE code also has to know whether or not a current VA.gov session exists, and if not, there is no reason to initiate an SSOe auto-logout.  Am I missing something that the BE code has the ability to determine in regrads to the SSOe session?

**Comment by patrickvinograd** [comment]
You're right that that will be triggered by the frontend, I agree there's no backend concern there.

**Comment by patrickvinograd** [comment]
So then the question I guess is - if the frontend sees that there is no SSOe session - can it tell whether that's due to the SSOe session expiring (in which case it should trigger logout), vs. the user not having had the feature flag enabled (in which case it should not trigger logout)? 

I believe we are pushing an `ssoe` indicator into the user profile, so frontend could use that.
Or, frontend could trigger the request either way and the backend could choose to ignore it if it sees that the user is not logged in via SSOe.

**Comment by ericbuckley** [comment]
I think the FE would need to check for 3 conditions, if all of them are true it would auto logout the user

1) `/keepalive` endpoint returns false (no SSOe session)
2) the `ssoe` [feature flag](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/user/authentication/selectors.js#L4) is enabled
3) `/v0/user` returns a 200 (the user has an active VA.gov session)

@patrickvinograd / @hinzed1127 does that sound right?

**Comment by hinzed1127** [comment]
That sounds accurate. Right now the logic I have for auto login checks for the things you called out, @ericbuckley : 
```js
ssoKeepAliveSession(); // this calls the keep-alive endpoint and updates localStorage values related to SSOe
const canCallSSO = this.props.useSSOe && !environment.isLocalhost(); // useSSOe is the feature flag

if (hasSession()) {
  if (canCallSSO && !hasSessionSSO()) {
    autoLogout();
  }
  this.props.initializeProfile();
} else if (canCallSSO && hasSessionSSO()) {
  autoLogin();
}
```

This address these 2 cases, with the assumption that the feature flag is enabled:

1. If there's _no_ active vets-api but an active SSOe session, log in
2. If there's an active vets-api session but _no_ SSOe session, log out
